Allow for running conformance tests by users with reduced privileges

mgencur opened the pull request:
Fixes #11588 (the test namespaces part and instructions)

Beginning of slack discussion: https://knative.slack.com/archives/CA1DTGZ2N/p1624609372028500

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Allow for passing specific test namespaces names via flags
* Added overlay-test-namespace.yaml that removes the test namespaces from the installed yamls (they're supposed to be created by cluster-admin
* Updated readme file with instructions on how to run the Conformance tests as a project admin user


**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note

```

mgencur commented:
/retest

mgencur commented:
/retest

mgencur commented:
/retest

dprotaso commented:
Re-reading the issue I don't think this PR will accomplish what you're looking for.

`knative_setup` is installing _many_ privileged resources & components. These components are necessary for the system to function - ie. ingress & certificate providers etc.

Since you're deleting all namespaces in the config you're going to break the knative deployment.  I believe `kapp` will do a proper diff of things missing in the manifest and delete it.


Are you just looking to run the tests _without_ setting up/creating the cluster? That can be accomplished by passing `--run-tests` the e2e scripts I believe

mgencur commented:
@dprotaso My goal is to be able to run conformance tests against a pre-created cluster which already has Knative installed. I don't think the flag `--run-tests` will help because it's still setting up many things that are not necessary for the tests. In fact, it's enough for me to call this to install the minimum required test resources and then immediately run conformance tests:
```
run_ytt \
    -f test/config/ytt/lib \
    -f test/config/ytt/values.yaml \
    -f test/config/ytt/core/resources.yaml \
    -f test/config/ytt/overlay-test-namespace | kubectl apply -f -
```
OK, the overlay will delete all namespace including knative-serving :-/ So this needs to be changed.
The `knative_setup` function installs both core Knative compononents and test resources. Perhaps installing the minimum test resources could be extracted and re-used.

mgencur commented:
I think it would very well work for me to just provide a correct `test/config/ytt/overlay-test-namespace` which would only prevent creating the extra test namespaces and then have a couple of notes in the readme about running the conformance tests. Because it would probably be as simple as:
1) Run the command for applying `test/config/ytt/core/resources.yaml` and `test/config/ytt/overlay-test-namespace`
2) Run tests via `go_test_e2e -timeout=30m ./test/conformance/api/... ./test/conformance/runtime/... ...`

dprotaso commented:
> My goal is to be able to run conformance tests against a pre-created cluster which already has Knative installed. 

You need a test namespace in addition to the config map and secret for conformance to run tests.

These are unfortunately hardcoded at the moment:
https://github.com/knative/serving/blob/c75484eac76bb7d6cafc5a06b754b2e7cb1f015b/test/e2e_constants.go#L25-L32

mgencur commented:
@dprotaso What do you think about this? Move the creation of the ConfigMap and Secret directly to Golang code, it's integral part of the test suite anyway. Then running the tests would be as simple as:
1) Create the three namespace in advance by an admin user
2) Run Golang Conformance tests as a regular user (not admin)
(no running of ytt in this case)

Maybe I could go even further and pass the three test namespaces via flags (`test-namespace`, `alternative-test-namespace`, `tls-test-namespace` (and keep the default values as before if the flags are undefined). That would allow for the admin to create a few ad-hoc namespaces in advance for the regular user and he could use them.

dprotaso commented:
> Move the creation of the ConfigMap and Secret directly to Golang code

Unfortunately - our [API specification](https://github.com/knative/specs/blob/main/specs/serving/knative-api-specification-1.0.md) doesn't include ConfigMap API surface. Thus changing the test in such a way _could_ cause failures for implementations that would otherwise be conformant.

> Maybe I could go even further and pass the three test namespaces via flags 

Just note I believe the expectation is the config maps and secrets used for conformance exist in the `test-namespace`

mgencur commented:
Hi @dprotaso ! Thanks for your valuable feedback. I've implemented and tested these changes:
* allow for passing specific test namespaces names via flags
* added overlay-test-namespace.yaml that removes the test namespaces from the installed yamls (they're supposed to be created by cluster-admin
* updated readme file with instructions on how to run the Conformance tests as a project admin user

salaboy commented:
@dprotaso I ended up looking at this PR because conformance tests are not working with the previous instructions. I think we need to review and merge this PR as soon as possible. I've reviewed the changes in this PR and they look OK to me. I don't see nothing concerning and as the test in some way are broken now I would merge this PR as soon as possible.

dprotaso commented on test/README.md:
````diff
@@ -48,16 +48,61 @@ which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._
+1. The cluster admin creates three test namespaces names: `serving-tests`, `serving-tests-alt`, `tls`.
+1. The project admin installs minimum test resources:
+    ```bash
+    ytt -f test/config/ytt/lib \
````
we don't need the `lib` directory as you're not importing any of the functions

dprotaso commented on test/README.md:
```diff
@@ -48,16 +48,61 @@ which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._
+    ytt -f test/config/ytt/lib \
+        -f test/config/ytt/values.yaml \
+        -f test/config/ytt/core/resources.yaml \
+        -f test/config/ytt/overlay-test-namespace.yaml | kubectl apply -f -
```
It may be simpler to split the `resources.yaml` file so you don't need to apply your overlay here

`test-cluster-resources.yaml`
`test-resources.yaml`

dprotaso commented on test/README.md:
````diff
@@ -48,16 +48,61 @@ which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._
+The tests can be run in arbitrary test namespaces. When specific namespaces are used
+their names must be passed to the ytt command via the `--data-value` flag as follows:
+```bash
+    --data-value data.values.serving.namespaces.test.default=serving-tests \
+    --data-value data.values.serving.namespaces.test.alternative=serving-tests-alt \
+    --data-value data.values.serving.namespaces.test.tls=tls
````
We don't need the `data.values.` prefix

```suggestion
    --data-value serving.namespaces.test.default=serving-tests \
    --data-value serving.namespaces.test.alternative=serving-tests-alt \
    --data-value serving.namespaces.test.tls=tls
```

dprotaso commented on test/config/ytt/overlay-test-namespace.yaml:
```diff
@@ -0,0 +1,16 @@
+#@ load("@ytt:data", "data")
```
I'd call `overlay-test-namespace.yaml` - `remove-test-namespaces.yaml` to be clear what's happening with the overlay

dprotaso commented on test/README.md:
````diff
@@ -48,16 +48,61 @@ which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._
+
+The specific namespaces must also be passed to the Golang test suite:
+```bash
+   --test-namespace=serving-tests \
+   --alt-test-namespace=serving-tests-alt \
+   --tls-test-namespace=tls
````
we're using go's flag library so these should be a single `-`

dprotaso commented on test/README.md:
````diff
@@ -48,16 +48,61 @@ which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._
+1. The project admin then runs the conformance test suite using the `--disable-logstream` flag:
+    ```bash
+    go test -v -tags=e2e -count=1 \
+      --disable-logstream \
+      --kubeconfig $PROJECT_ADMIN_KUBECONFIG \
````
we're using go's flag library so these should be a single `-`

mgencur replied:
I believe it can be both variants, but I've changed that to `-kubeconfig=$PROJECT_ADMIN_KUBECONFIG`

mgencur replied:
https://golang.org/pkg/flag/#hdr-Command_line_flag_syntax , there are actually 6 options :-D
```
-flag
-flag=x
-flag x  // non-boolean flags only
```
And, one or two minus signs allowed.

dprotaso replied:
oh yeah - interesting
> One or two minus signs may be used; they are equivalent.

dprotaso commented on test/e2e_flags.go:
```diff
@@ -68,5 +71,14 @@ func initializeServingFlags() *ServingEnvironmentFlags {
 	flag.BoolVar(&f.DisableLogStream, "disable-logstream", false,
 		"Set this flag to disable streaming logs from system components")
 
+	flag.StringVar(&f.TestNamespace, "test-namespace", "serving-tests",
+		"Set this flag to change the default namespace for running tests.")
+
+	flag.StringVar(&f.AltTestNamespace, "alt-test-namespace", "serving-tests-alt",
+		"Set this flag to change the alternative namespace for running tests.")
+
+	flag.StringVar(&f.TLSTestNamespace, "tls-test-namespace", "tls",
+		"Set this flag to change the namespace for running TLS tests.")
+
```
Can we preserve the original defaults:

`serving-tests`
`serving-tests-alt`
`tls`

mgencur replied:
They're preserved. Or no? The default values are `serving-tests`, `serving-tests-alt`, `tls` . Am I missing something?

dprotaso replied:
sorry I misread the flag name as default value

dprotaso commented:
/lgtm
/approve
